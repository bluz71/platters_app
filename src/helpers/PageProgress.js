import NProgress from 'nprogress'

class PageProgress {
  constructor () {
    this.loaded = false
  }

  start () {
    if (!this.loaded) {
      NProgress.start()
    }
  }

  done () {
    if (!this.loaded) {
      this.loaded = true
      NProgress.done()
    }

    return true
  }
}

export default PageProgress
