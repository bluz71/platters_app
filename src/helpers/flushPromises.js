// Details:
//   https://blog.rescale.com/testing-promise-side-effects-with-asyncawait
//   https://github.com/facebook/jest/issues/2157#issuecomment-279171856

const flushPromises = () => new Promise(resolve => setImmediate(resolve));

export default flushPromises;
