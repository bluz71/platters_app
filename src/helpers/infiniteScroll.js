// Code snippet based on:
//   http://blog.sodhanalibrary.com/2016/08/detect-when-user-scrolls-to-bottom-of.html
//
const infiniteScoll = (pagination, onPageEnd) => {
  const moreToScroll = pagination && pagination.next_page;
  if (!moreToScroll) {
    return;
  }

  const windowHeight = window.innerHeight;
  const body = document.body;
  const html = document.documentElement;
  const docHeight =
    Math.max(
      body.scrollHeight,
      body.offsetHeight,
      html.clientHeight,
      html.scrollHeight,
      html.offsetHeight
    ) - 20;
  const windowBottom = windowHeight + window.pageYOffset;

  if (windowBottom >= docHeight) {
    onPageEnd();
  }
};

export default infiniteScoll;
