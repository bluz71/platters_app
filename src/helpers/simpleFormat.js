// Code snippet provided by:
//   https://makandracards.com/makandra/1395-simple_format-helper-for-javascript
//
const simpleFormat = str => {
  str = str.replace(/\r\n?/, '\n')
  str = str.trim()

  if (str.length > 0) {
    str = str.replace(/\n\n+/g, '</p><p>')
    str = str.replace(/\n/g, '<br />')
    str = `<p>${str}</p>`
  }

  return str
}

export default simpleFormat
