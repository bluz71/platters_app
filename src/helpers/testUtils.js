import { appAuth } from '../lib/appAuth';

// Details:
//   https://blog.rescale.com/testing-promise-side-effects-with-asyncawait
//   https://github.com/facebook/jest/issues/2157#issuecomment-279171856
const flushPromises = () => new Promise(resolve => setImmediate(resolve));

const logInUser = () => (
  appAuth.logIn('eyJhbGciOiJIUzI1NiJ9.eyJ1c2VyIjo1MywibmFtZSI6ImRlbm5pc2I1NSI\
sImFkbWluIjpmYWxzZSwiZXhwIjoxNTM3ODQ4NDA0LCJpc3MiOiJwbGF0dGVycyIsImF1ZCI6InBs\
YXR0ZXJzX2FwcCJ9.ZUV69wWVXfiiMtQ2ZFPvgqvhZWwQ4WC1bI15Q9DpFCU')
);

const logOutUser = () => (
  appAuth.logOut()
);

export { flushPromises, logInUser, logOutUser };
