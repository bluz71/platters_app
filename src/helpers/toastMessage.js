import { toast } from 'react-toastify';

export const toastAlert = (message) => {
  toast.error(message, { className: 'ToastAlert' });
};

export const toastNotice = (message) => {
  toast.success(message, { className: 'ToastNotice' });
};
