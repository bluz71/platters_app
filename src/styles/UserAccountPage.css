.UserAccount .sub-header h1 {
  font-size: 2em;
}

.UserAccount .avatar img {
  height: 125px;
  width: 125px;
  border: 1px solid #eee;
  border-radius: 4px;
}

.UserAccount .avatar img:hover {
  border: 1px solid #3c9a5f;
}
