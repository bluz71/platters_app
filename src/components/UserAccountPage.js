import React, { Component } from 'react'
import axios from 'axios'
import { Redirect } from 'react-router-dom'
import {
  Row,
  Col,
  PageHeader,
  Well,
  Form,
  FormGroup,
  ControlLabel,
  FormControl,
  HelpBlock,
  Button
} from 'react-bootstrap'
import FontAwesome from 'react-fontawesome'
import md5 from 'md5'
import { API_HOST } from '../config'
import { appAuth } from '../lib/appAuth'
import { toastAlert, toastNotice } from '../helpers/toastMessage'
import '../styles/UserAccountPage.css'

class UserAccountPage extends Component {
  constructor (props) {
    super(props)

    this.userSlug = props.match.params.id
    this.userUserEndPoint = `${API_HOST}/api/users/${this.userSlug}`

    this.user = appAuth.currentUser()
    document.title = `Platters App - ${this.userSlug}`

    this.state = {
      updateButtonText: 'Update',
      nameErrors: []
    }

    // Bind 'this' for callback functions.
    this.handleUserUpdate = this.handleUserUpdate.bind(this)
    this.handleDeleteAccount = this.handleDeleteAccount.bind(this)
    this.handleNameBlur = this.handleNameBlur.bind(this)
  }

  handleUserUpdate (event) {
    event.preventDefault()

    this.setState({
      updateButtonText: (
        <div>
          <FontAwesome name='spinner' spin pulse /> Updating...
        </div>
      )
    })

    const userUpdate = {
      user: {
        name: this.nameInput.value,
        password: this.passwordInput.value
      }
    }

    this.putUserUpdate(userUpdate)
  }

  handleDeleteAccount () {
    if (!window.confirm('Are you sure you want to delete your account?')) {
      return
    }

    this.deleteUser()
  }

  handleNameBlur () {
    const name = this.nameInput.value
    if (name.length < 4 || name.length > 20) {
      this.setState({
        nameErrors: ['Account name must be between 4 and 20 characters long']
      })
    } else {
      this.setState({ nameErrors: [] })
    }
  }

  async putUserUpdate (userUpdate) {
    axios
      .put(this.userUserEndPoint, userUpdate, await appAuth.headers())
      .then(response => {
        this.setState({ updateButtonText: 'Update' })
        appAuth.logIn(response.data.auth_token)
        toastNotice('Your account has been successfully updated')
        this.props.history.push('/')
      })
      .catch(error => {
        this.setState({ updateButtonText: 'Update' })
        if (error.response && error.response.status === 400) {
          toastAlert('You can only update your own account')
        } else if (error.response && error.response.status === 406) {
          toastAlert('Your account could not be updated')
          this.setState({
            nameErrors: error.response.data.errors
          })
        } else if (error.response && error.response.status === 404) {
          toastAlert(`The user '${this.userSlug}' does not exist`)
          this.props.history.push('/password/new')
        } else {
          toastAlert('Server error, please try again later')
        }
      })
  }

  async deleteUser () {
    axios
      .delete(this.userUserEndPoint, await appAuth.headers())
      .then(response => {
        appAuth.logOut()
        toastNotice('Your account has been successfully deleted')
        this.props.history.push('/')
      })
      .catch(error => {
        if (error.response && error.response.status === 400) {
          toastAlert('You can only delete your own account')
        } else if (error.response && error.response.status === 404) {
          toastAlert(`The user '${this.userSlug}' does not exist`)
          this.props.history.push('/password/new')
        } else {
          toastAlert('Server error, please try again later')
        }
      })
  }

  gravatarURL () {
    const hash = md5(this.user.email.trim().toLowerCase())

    return `https://gravatar.com/avatar/${hash}?s=160&r=pg&d=identicon`
  }

  renderNameErrors () {
    return this.state.nameErrors.map((error, index) => (
      <li key={index} className='list-group-item list-group-item-danger'>
        {error}
      </li>
    ))
  }

  renderAccount () {
    return (
      <div>
        <PageHeader>
          Account <small>{this.user.email}</small>
        </PageHeader>
        <Well>
          <Form horizontal onSubmit={this.handleUserUpdate}>
            <ul className='list-group'>{this.renderNameErrors()}</ul>

            <FormGroup>
              <Col componentClass={ControlLabel} md={2}>
                Name
              </Col>
              <Col md={9}>
                <FormControl
                  type='text'
                  defaultValue={this.user.name}
                  className='name'
                  inputRef={input => (this.nameInput = input)}
                  onBlur={this.handleNameBlur}
                />
              </Col>
            </FormGroup>

            <FormGroup>
              <Col componentClass={ControlLabel} md={2}>
                Password
              </Col>
              <Col md={9}>
                <FormControl
                  type='password'
                  className='password'
                  inputRef={input => (this.passwordInput = input)}
                />
                <HelpBlock>
                  You must enter a password even if you are only changing your
                  account name
                </HelpBlock>
              </Col>
            </FormGroup>

            <FormGroup>
              <Col mdOffset={2} md={9}>
                <Button
                  type='submit'
                  bsStyle='success'
                  bsSize='small'
                  className='account-update'
                >
                  {this.state.updateButtonText}
                </Button>
              </Col>
            </FormGroup>
          </Form>
        </Well>
      </div>
    )
  }

  renderAvatar () {
    return (
      <div className='sub-header avatar'>
        <PageHeader>Avatar</PageHeader>
        <Well>
          <p>
            Set a custom avatar for Platters by registering at Gravatar with
            your email address.
          </p>

          <p>
            Gravatar is a global free avatar service used by many applications
            including: Wordpress, Github and Stackoverflow. Click on the avatar
            below to be taken to Gravatar.
          </p>

          <a href='https://gravatar.com'>
            <img
              className='img-responsive'
              src={this.gravatarURL()}
              alt={this.user.name}
            />
          </a>
        </Well>
      </div>
    )
  }

  renderDeleteAccount () {
    return (
      <div className='sub-header'>
        <PageHeader>Delete account</PageHeader>
        <Well>
          <p>
            Once you delete your account all your comments will also be deleted,
            there is no going back. Please be certain!
          </p>
          <Button
            type='submit'
            bsStyle='danger'
            bsSize='small'
            className='account-delete'
            onClick={this.handleDeleteAccount}
          >
            Delete account
          </Button>
        </Well>
      </div>
    )
  }

  render () {
    if (!this.user || this.userSlug !== this.user.slug) {
      toastAlert(
        `Can not display '${
          this.userSlug
        }' user account, you may only access your own account`
      )
      return <Redirect to='/' />
    }

    return (
      <Row>
        <Col md={10} mdOffset={1} className='UserAccount'>
          {this.renderAccount()}
          {this.renderAvatar()}
          {this.renderDeleteAccount()}
        </Col>
      </Row>
    )
  }
}

export default UserAccountPage
