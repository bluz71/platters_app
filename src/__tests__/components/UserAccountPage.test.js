import React from 'react'
import { shallow, mount } from 'enzyme'
import { MemoryRouter } from 'react-router-dom'
import { flushPromises, logInUser } from '../../helpers/testUtils'
import * as toastModule from '../../helpers/toastMessage'
import UserAccountPage from '../../components/UserAccountPage'

describe('<UserAccountPage />', () => {
  it('renders without crashing', () => {
    shallow(
      <MemoryRouter initialEntries={[{ key: 'testKey' }]}>
        <UserAccountPage />
      </MemoryRouter>
    )
  })

  it('renders user account page', () => {
    logInUser()
    const match = { params: { id: 'fred' } }
    const wrapper = mount(
      <MemoryRouter initialEntries={[{ key: 'testKey' }]}>
        <UserAccountPage match={match} />
      </MemoryRouter>
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('update user with valid name and password', async () => {
    logInUser()
    const spyNotice = jest.spyOn(toastModule, 'toastNotice')
    const match = { params: { id: 'fred' } }
    const wrapper = mount(
      <MemoryRouter initialEntries={[{ key: 'testKey' }]}>
        <UserAccountPage match={match} />
      </MemoryRouter>
    )
    wrapper.find('input.name').instance().value = 'freddy'
    wrapper.find('input.password').instance().value = 'password9'
    wrapper.find('button.account-update').simulate('submit')
    await flushPromises()
    expect(spyNotice).toHaveBeenCalledWith(
      'Your account has been successfully updated'
    )
    spyNotice.mockRestore()
  })

  it('update user with invalid details', async () => {
    logInUser()
    const spyAlert = jest.spyOn(toastModule, 'toastAlert')
    const match = { params: { id: 'fred' } }
    const wrapper = mount(
      <MemoryRouter initialEntries={[{ key: 'testKey' }]}>
        <UserAccountPage match={match} />
      </MemoryRouter>
    )
    wrapper.find('input.name').instance().value = 'fre'
    wrapper.find('input.password').instance().value = 'pass'
    wrapper.find('button.account-update').simulate('submit')
    await flushPromises()
    expect(spyAlert).toHaveBeenCalledWith('Your account could not be updated')
    spyAlert.mockRestore()
  })

  it('delete user', async () => {
    logInUser()
    // Mock window.confirm which will be invoked when deleting a user.
    window.confirm = jest.fn(() => true)
    const spyNotice = jest.spyOn(toastModule, 'toastNotice')
    const match = { params: { id: 'fred' } }
    const wrapper = mount(
      <MemoryRouter initialEntries={[{ key: 'testKey' }]}>
        <UserAccountPage match={match} />
      </MemoryRouter>
    )
    wrapper.find('button.account-delete').simulate('click')
    await flushPromises()
    expect(spyNotice).toHaveBeenCalledWith(
      'Your account has been successfully deleted'
    )
    spyNotice.mockRestore()
  })
})
