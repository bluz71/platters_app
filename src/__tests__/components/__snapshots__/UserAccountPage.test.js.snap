// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserAccountPage /> renders user account page 1`] = `
<MemoryRouter
  initialEntries={
    Array [
      Object {
        "key": "testKey",
      },
    ]
  }
>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "key": "testKey",
            "pathname": "/",
            "search": "",
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "key": "testKey",
          "pathname": "/",
          "search": "",
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <UserAccountPage
      match={
        Object {
          "params": Object {
            "id": "fred",
          },
        }
      }
    >
      <Row
        bsClass="row"
        componentClass="div"
      >
        <div
          className="row"
        >
          <Col
            bsClass="col"
            className="UserAccount"
            componentClass="div"
            md={10}
            mdOffset={1}
          >
            <div
              className="UserAccount col-md-10 col-md-offset-1"
            >
              <div>
                <PageHeader
                  bsClass="page-header"
                >
                  <div
                    className="page-header"
                  >
                    <h1>
                      Account 
                      <small>
                        fred@example.com
                      </small>
                    </h1>
                  </div>
                </PageHeader>
                <Well
                  bsClass="well"
                >
                  <div
                    className="well"
                  >
                    <Form
                      bsClass="form"
                      componentClass="form"
                      horizontal={true}
                      inline={false}
                      onSubmit={[Function]}
                    >
                      <form
                        className="form-horizontal"
                        onSubmit={[Function]}
                      >
                        <ul
                          className="list-group"
                        />
                        <FormGroup
                          bsClass="form-group"
                        >
                          <div
                            className="form-group"
                          >
                            <Col
                              bsClass="col"
                              componentClass={[Function]}
                              md={2}
                            >
                              <ControlLabel
                                bsClass="control-label"
                                className="col-md-2"
                                srOnly={false}
                              >
                                <label
                                  className="col-md-2 control-label"
                                >
                                  Name
                                </label>
                              </ControlLabel>
                            </Col>
                            <Col
                              bsClass="col"
                              componentClass="div"
                              md={9}
                            >
                              <div
                                className="col-md-9"
                              >
                                <FormControl
                                  bsClass="form-control"
                                  className="name"
                                  componentClass="input"
                                  defaultValue="fred"
                                  inputRef={[Function]}
                                  onBlur={[Function]}
                                  type="text"
                                >
                                  <input
                                    className="name form-control"
                                    defaultValue="fred"
                                    onBlur={[Function]}
                                    type="text"
                                  />
                                </FormControl>
                              </div>
                            </Col>
                          </div>
                        </FormGroup>
                        <FormGroup
                          bsClass="form-group"
                        >
                          <div
                            className="form-group"
                          >
                            <Col
                              bsClass="col"
                              componentClass={[Function]}
                              md={2}
                            >
                              <ControlLabel
                                bsClass="control-label"
                                className="col-md-2"
                                srOnly={false}
                              >
                                <label
                                  className="col-md-2 control-label"
                                >
                                  Password
                                </label>
                              </ControlLabel>
                            </Col>
                            <Col
                              bsClass="col"
                              componentClass="div"
                              md={9}
                            >
                              <div
                                className="col-md-9"
                              >
                                <FormControl
                                  bsClass="form-control"
                                  className="password"
                                  componentClass="input"
                                  inputRef={[Function]}
                                  type="password"
                                >
                                  <input
                                    className="password form-control"
                                    type="password"
                                  />
                                </FormControl>
                                <HelpBlock
                                  bsClass="help-block"
                                >
                                  <span
                                    className="help-block"
                                  >
                                    You must enter a password even if you are only changing your account name
                                  </span>
                                </HelpBlock>
                              </div>
                            </Col>
                          </div>
                        </FormGroup>
                        <FormGroup
                          bsClass="form-group"
                        >
                          <div
                            className="form-group"
                          >
                            <Col
                              bsClass="col"
                              componentClass="div"
                              md={9}
                              mdOffset={2}
                            >
                              <div
                                className="col-md-9 col-md-offset-2"
                              >
                                <Button
                                  active={false}
                                  block={false}
                                  bsClass="btn"
                                  bsSize="small"
                                  bsStyle="success"
                                  className="account-update"
                                  disabled={false}
                                  type="submit"
                                >
                                  <button
                                    className="account-update btn btn-sm btn-success"
                                    disabled={false}
                                    type="submit"
                                  >
                                    Update
                                  </button>
                                </Button>
                              </div>
                            </Col>
                          </div>
                        </FormGroup>
                      </form>
                    </Form>
                  </div>
                </Well>
              </div>
              <div
                className="sub-header avatar"
              >
                <PageHeader
                  bsClass="page-header"
                >
                  <div
                    className="page-header"
                  >
                    <h1>
                      Avatar
                    </h1>
                  </div>
                </PageHeader>
                <Well
                  bsClass="well"
                >
                  <div
                    className="well"
                  >
                    <p>
                      Set a custom avatar for Platters by registering at Gravatar with your email address.
                    </p>
                    <p>
                      Gravatar is a global free avatar service used by many applications including: Wordpress, Github and Stackoverflow. Click on the avatar below to be taken to Gravatar.
                    </p>
                    <a
                      href="https://gravatar.com"
                    >
                      <img
                        alt="fred"
                        className="img-responsive"
                        src="https://gravatar.com/avatar/6255165076a5e31273cbda50bb9f9636?s=160&r=pg&d=identicon"
                      />
                    </a>
                  </div>
                </Well>
              </div>
              <div
                className="sub-header"
              >
                <PageHeader
                  bsClass="page-header"
                >
                  <div
                    className="page-header"
                  >
                    <h1>
                      Delete account
                    </h1>
                  </div>
                </PageHeader>
                <Well
                  bsClass="well"
                >
                  <div
                    className="well"
                  >
                    <p>
                      Once you delete your account all your comments will also be deleted, there is no going back. Please be certain!
                    </p>
                    <Button
                      active={false}
                      block={false}
                      bsClass="btn"
                      bsSize="small"
                      bsStyle="danger"
                      className="account-delete"
                      disabled={false}
                      onClick={[Function]}
                      type="submit"
                    >
                      <button
                        className="account-delete btn btn-sm btn-danger"
                        disabled={false}
                        onClick={[Function]}
                        type="submit"
                      >
                        Delete account
                      </button>
                    </Button>
                  </div>
                </Well>
              </div>
            </div>
          </Col>
        </div>
      </Row>
    </UserAccountPage>
  </Router>
</MemoryRouter>
`;
